### PR TITLE
RequiresAssemblyFilesAttribute resource manager changes

### PR DESCRIPTION
**Pull request by tlakollo**
PR for #1786
-Change resource file to have a default message and accept arguments
-Add test for RequiresAssemblyFilesAttribute with Url named argument only

### CODE REVIEW

File: src/ILLink.RoslynAnalyzer/Resources.resx
Comment (agocke) [inline src/ILLink.RoslynAnalyzer/Resources.resx:142]: I'm not sure `bundling the application code` is a well-understood term. The above terminology uses "embedded in a single-file app," which seems understandable for things like Xamarin or Blazor.

File: src/ILLink.RoslynAnalyzer/Resources.resx
Comment (agocke) [inline src/ILLink.RoslynAnalyzer/Resources.resx:139]: I don't think this is supposed to be the title of the analyzer, it's the title of the warning, which is like a short description of what the warning is supposed to prevent.

Something like, "Avoid calling members marked with `RequiresAssemblyFilesAttribute` when publishing as a single file"?

File: test/ILLink.RoslynAnalyzer.Tests/RequiresAssemblyFilesAnalyzerTests.cs
Comment (agocke) [inline test/ILLink.RoslynAnalyzer.Tests/RequiresAssemblyFilesAnalyzerTests.cs:50]: These tests should print out a comment with the message when the test fails. If you include the comment, you can see if it looks right in the message and make the code easier to review.

Comment (tlakollo) [reply test/ILLink.RoslynAnalyzer.Tests/RequiresAssemblyFilesAnalyzerTests.cs:50]: Are we sure we want the comments? they got deleted from the RequiresUnreferencedCodeAnalyzerTests too. I'm fine with or without them just want to homologate things between the analyzers code.

File: src/ILLink.RoslynAnalyzer/Resources.resx
Comment (mateoatr) [inline src/ILLink.RoslynAnalyzer/Resources.resx:139]: Nit: can we change (here and above) the backticks for common apostrophe (') similar to

https://github.com/mono/linker/blob/3c3771c5f0818f9b88387bd2bc1d9e58712e6b31/src/ILLink.RoslynAnalyzer/Resources.resx#L124

Or https://github.com/mono/linker/blob/master/src/linker/Resources/Strings.resx ?

File: src/ILLink.RoslynAnalyzer/Resources.resx
Comment (mateoatr) [inline src/ILLink.RoslynAnalyzer/Resources.resx:121]:
```suggestion
    <value>Avoid calling methods marked with 'RequiresUnreferencedCodeAttribute' when trimming application code</value>
```

Also I'm not sure that something marked with this attribute is something that the user should avoid calling, it's just a method that requires dynamic access, which is dangerous when linking the application (but I really don't have a good suggestion for framing this title).

File: src/ILLink.RoslynAnalyzer/Resources.resx
Comment (mateoatr) [inline src/ILLink.RoslynAnalyzer/Resources.resx:124]: Since you're editing things here, I think we should update this message to match the one we produce from within the linker: https://github.com/mono/linker/pull/1778#discussion_r564429652

File: src/ILLink.RoslynAnalyzer/Resources.resx
Comment (marek-safar) [reply src/ILLink.RoslynAnalyzer/Resources.resx:139]: Is calling the right terminology here?. Are we going to add a new message for each bundle like format?

Comment (tlakollo) [reply src/ILLink.RoslynAnalyzer/Resources.resx:139]: > Is calling the right terminology here?

I'm changing the message to match something similar to https://github.com/mono/linker/pull/1778#discussion_r564429652
> Are we going to add a new message for each bundle like format?

Since they all will use the same custom attribute I would expect to print the same message independent of the kind of bundle

File: test/ILLink.RoslynAnalyzer.Tests/TestCaseUtils.cs
Comment (agocke) [inline test/ILLink.RoslynAnalyzer.Tests/TestCaseUtils.cs:69]: It looks like this no longer handles `params` arrays properly, namely it only looks at the second argument, instead of all the arguments.

File: src/ILLink.RoslynAnalyzer/Resources.resx
Comment (agocke) [reply src/ILLink.RoslynAnalyzer/Resources.resx]: We could rephrase by saying something like "Members annotated with `RequiresUnreferencedCodeAttribute` may behave differently or fail if the unreferenced code is removed by trimming"

I think that's a decent description of the problem, but it doesn't really give a call to action, like "don't call this when trimming." I don't have a strong opinion either way, but I think we should decide on what kind of title we want to present.

The context for this resource is that when you're looking at a list of the warnings, this will be the description of what the warning is about. The message is what's actually printed when the warning fires.

Comment (tlakollo) [reply src/ILLink.RoslynAnalyzer/Resources.resx]: In a previous commit, I changed the message to:
Methods marked with 'RequiresUnreferencedCodeAttribute' require dynamic access otherwise can break functionality when trimming application code

File: test/Mono.Linker.Tests.Cases/RequiresCapability/RequiresUnreferencedCodeCapability.cs
Comment (agocke) [inline test/Mono.Linker.Tests.Cases/RequiresCapability/RequiresUnreferencedCodeCapability.cs:59]: Why add `new string[]`?

**Review by agocke — APPROVED** [review]
LGTM, although I'd prefer if we settle on one style for params array invocations in `ExpectedWarning`: either use `new string[]` everywhere, or don't use it anywhere.

**Review by mateoatr — CHANGES_REQUESTED** [review]
Can we remove the `new string[]` usage in the tests? Why do we need it? Looks to me that it is enough to just look for the warning code and look for the message that was specified in the `Message` parameter of the method that we think is the origin of the warning.

File: test/Mono.Linker.Tests.Cases/RequiresCapability/RequiresUnreferencedCodeCapability.cs
Comment (tlakollo) [reply test/Mono.Linker.Tests.Cases/RequiresCapability/RequiresUnreferencedCodeCapability.cs]: Ok, I changed the code to not support array creation, the code will throw in GetStringFromExpr if it ever tries to be used

File: test/Mono.Linker.Tests.Cases/RequiresCapability/RequiresUnreferencedCodeCapability.cs
Comment (mateoatr) [inline test/Mono.Linker.Tests.Cases/RequiresCapability/RequiresUnreferencedCodeCapability.cs:111]: It's enough to only validate the message that's coming from the annotated method. We explicitly removed full text validation previously: https://github.com/mono/linker/pull/1778#discussion_r564432966

Comment (tlakollo) [reply test/Mono.Linker.Tests.Cases/RequiresCapability/RequiresUnreferencedCodeCapability.cs:111]: My line of thinking was that this is not full-text validation, I only validate the method name and the message which are things that should not change in the message. The linker has code in the result checker to check for the method with the ExpectedWarning attribute. The analyzer on the other hand will match any messages that are within the span, which means that there is a chance for the analyzer to check for messages that were not produced by the method marked with ExpectedWarning. Although in this file, no message is repeated, so using the message is enough. I will change the implementation.

**Pull request merged**
